Optimize load and unload operations in vineyard LLM cache

DwyaneShi opened the pull request:
- optimize pin memory buffer to make the kv tensors of a chunk sit in a
  contiguous memory region
- using a cuda staging buffer to avoid the inefficiency of non-contiguous
HBM->DRAM / DRAM->HBM memcpy

FILL IN THE PR DESCRIPTION HERE

FIX #xxxx (*link existing issues this PR will resolve*)

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>
<!-- inside this <details> section, markdown rendering does not work, so we use raw html here. -->
<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Adding or changing kernels</h3>
<p>Each custom kernel needs a schema and one or more implementations to be registered with PyTorch.</p>
<ul>
    <li>Make sure custom ops are registered following PyTorch guidelines: <a href="https://pytorch.org/tutorials/advanced/cpp_custom_ops.html#cpp-custom-ops-tutorial">Custom C++ and CUDA Operators</a> and <a href="https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU">The Custom Operators Manual</a></li>
    <li>Custom operations that return <code>Tensors</code> require meta-functions. Meta-functions should be implemented and registered in python so that dynamic dims can be handled automatically. See above documents for a description of meta-functions.</li>
    <li>Use <a href="https://pytorch.org/docs/stable/library.html#torch.library.opcheck"><code>torch.libary.opcheck()</code></a> to test the function registration and meta-function for any registered ops.  See <code>tests/kernels</code> for examples.</li>
    <li>When changing the C++ signature of an existing op, the schema must be updated to reflect the changes.</li>
    <li>If a new custom type is needed, see the following document: <a href="https://docs.google.com/document/d/18fBMPuOJ0fY5ZQ6YyrHUppw9FA332CpNtgB6SOIgyuA">Custom Class Support in PT2</a>.
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>


</details>

Jeffwan commented:
It's a little bit busy earlier this week. I will help review this change later today.

Jeffwan commented on vllm/worker/vineyard_llm_cache.py:
```diff
@@ -369,9 +384,11 @@ def update_seq_kv_caches(
-            self.buffer[:, j, :update_token_size].copy_(
+            self.cuda_buffer[:, j, :update_token_size].copy_(
                 kv_caches[j][:, slot_mapping // block_size, slot_mapping % block_size])
+        self.buffer.copy_(self.cuda_buffer)
```
I kind of forget why do we copy from `buffer` to `cuda_buffer` here?

DwyaneShi replied:
this is a copy from `cuda_buffer` to `buffer`, which converts original non-contiguous HBM->DRAM copy (kv_caches (non contiguous) -> buffer) to contiguous copy (kv_caches (non contiguous) -> cuda_buffer (contiguous) -> buffer)

Jeffwan reviewed (approved):
overall change looks good to me

FuturisticWater commented on vllm/worker/vineyard_llm_cache.py:
```diff
@@ -383,9 +384,11 @@ def update_seq_kv_caches(
         start_unload = torch.cuda.Event(enable_timing=True)
         end_unload = torch.cuda.Event(enable_timing=True)
         start_unload.record()
+        # using a cuda staging buffer to avoid the inefficiency of non-contiguous HBM->DRAM memcpy
```
Could you please document the use of the cuda staging buffer in PR description?

DwyaneShi replied:
updated

FuturisticWater commented on vllm/worker/vineyard_llm_cache.py:
```diff
@@ -242,7 +243,7 @@ def prefetch_seq_kv_caches(
         # approach reduces the number of transfers between CPU and GPU,
         # leading to faster overall performance.
-        buffer = self.buffer.cuda()[:, :, :matched]
+        buffer = self.cuda_buffer.copy_(self.buffer)[:, :, :matched]
```
What's the purpose of this change? This doesn't seem to be related to unload?

DwyaneShi replied:
this is the optimization for `load`. copying the whole `buffer` to HBM is a contiguous copying and thus more efficient than legacy impl.

DwyaneShi replied:
removing `.cuda()` is because we have introduced `cuda_buffer` for optimizing `unload` and thus we don't need to extra allocation but reusing this `cuda_buffer` for `load` as well. `.cuda()` involves an implicit allocation.

FuturisticWater replied:
So we also applied the contiguous memory copy optimization to load as well? If so, please update the commit/PR description to reflect this.

Also, if Le is only waiting for this memory copy optimization, then I suggest we merge this commit in a separate PR, since it's much smaller than the commit that eliminates the extra re-computations.

DwyaneShi replied:
1. It's already in the description. We have gone through the benefits of the optimization for load in the meeting. The second commit for unload is a follow-up optimization we figured out recently.
2. Actually most of the changes are for load and unload. Changes for skipping re-compute are much smaller and straight forward. I will split the PR later.

DwyaneShi replied:
changes for skipping re-compute have been moved into https://github.com/aibrix/vllm/pull/10

FuturisticWater replied:
Now this PR is self-contained and easier to understand: previously I got the impression that this piece of comments are already there and was wondering why it was not effective and need further tweaks. Now I see all of these changes are part of the optimization of this PR.

FuturisticWater commented on vllm/worker/vineyard_llm_cache.py:
```diff
@@ -148,7 +158,7 @@ def prefetch_seq_kv_caches(
 
             # leave at least one token unmatched
-            token_chunk_size -= 1
+            # token_chunk_size -= 1
```
This seems to related to skipping the re-computation?

FuturisticWater commented on vllm/worker/vineyard_llm_cache.py:
```diff
@@ -188,6 +198,9 @@ def prefetch_seq_kv_caches(
+        # no need to minus 1 one more time. matched = min(matched, token_chunk_size - 1)
+        if seq_group_metadata is not None and seq_group_metadata.is_sampling_enabled:
+            matched = min(matched, token_chunk_size - 1)
         # synchronized across tensor parallel ranks
```
This also seems to related to skipping the re-computation?

FuturisticWater commented on vllm/worker/vineyard_llm_cache.py:
```diff
@@ -198,9 +211,6 @@ def prefetch_seq_kv_caches(
-        # we force to use token_chunk_size - 1 to trigger KV recomputation
-        # TODO: this should be revisited later. We are looking for solutions to fully avoid computation.
-        matched = min(matched, token_chunk_size - 1)
         if matched <= 0:
```
same here

Commit:
Optimize load and unload operations in vineyard LLM cache

- optimize pin memory buffer to make the kv tensors of a chunk sit in a
  contiguous memory region
- using a cuda staging buffer to avoid the inefficiency of non-contiguous
HBM->DRAM / DRAM->HBM memcpy

Signed-off-by: Haiyang Shi <haiyang.shi@bytedance.com>
